models/AmazonS3.cfc

```diff
@@ -261,21 +264,30 @@ component accessors="true" singleton {
 	/**
 	 * This function builds variables.UrlEndpoint and variables.URLEndpointHostname according to credentials and ssl configuration, usually called after init() for you automatically.
 	 */
-	AmazonS3 function buildUrlEndpoint(){
+	AmazonS3 function buildUrlEndpoint( string bucketName ){
 		// Build accordingly
 		var URLEndPointProtocol = ( variables.ssl ) ? "https://" : "http://";
 
 		var hostnameComponents = [];
-		if ( variables.awsDomain contains "amazonaws.com" ) {
+		if ( variables.urlStyle == "path" ) {
+			if ( variables.awsDomain contains "amazonaws.com" ) {
+				hostnameComponents.append( "s3" );
+			}
+			if ( len( variables.awsRegion ) ) {
+				hostnameComponents.append( variables.awsRegion );
+			}
+		} else if ( variables.urlStyle == "virtual" ) {
+			if ( !isNull( arguments.bucketName ) ) {
```

Comment:
Add a check for whether `variables.awsDomain contains "amazonaws.com"` here and below.  Then this could be used with a cloudfront distribution as well.

Reply:
Are these in the correct place?

Reply:
Not quite.   If the hostname doesn't contain `amazonaws.com` we don't want to modify the hostname at all.  In a cloudfront distribution or a reverse proxy, for example, the bucket _is_ the path.  So we wouldn't want to execute any modifications if it's not an AWS domain.

Reply:
If you want to test this out, to see what I mean, go in to the AWS control panel, go to "CloudFront" and create a distribution on the bucket.  Then use that as the `awsDomain` configuration setting.

Reply:
I think that will do it. Sorry for the back and forth but I appreciate your patiance.